diagnostics:  "try using a conversion method: `*i.to_string().to_string()"

Example I came up with (tried to reduce from a larger set of code I found this in)
````rust
fn main() {
	let mut v = vec!["hello", "this", "is", "a", "test"];

	let v2 = Vec::new();
	
	v.into_iter().map(|s|s.to_owned()).collect::<Vec<_>>();
	
	let mut a = String::new();
	for i in v {
		a = *i.to_string();
		v2.push(a);
	}
}
````
This gives a rather confusing error message:
````
error[E0308]: mismatched types
  --> src/main.rs:10:7
   |
10 |         a = *i.to_string();
   |             ^^^^^^^^^^^^^^
   |             |
   |             expected struct `std::string::String`, found str
   |             help: try using a conversion method: `*i.to_string().to_string()`
   |
   = note: expected type `std::string::String`
              found type `str`
````

meta:
````
rustc 1.30.0-nightly (d5a448b3f 2018-08-13)
binary: rustc
commit-hash: d5a448b3f47b22c9cb010198bdcc49d4392b090b
commit-date: 2018-08-13
host: x86_64-unknown-linux-gnu
release: 1.30.0-nightly
LLVM version: 7.0
````

--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -4629,8 +4629,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             let methods = self.get_conversion_methods(expr.span, expected, found);
             if let Ok(expr_text) = self.sess().codemap().span_to_snippet(expr.span) {
                 let suggestions = iter::repeat(expr_text).zip(methods.iter())
-                    .map(|(receiver, method)| format!("{}.{}()", receiver, method.ident))
-                    .collect::<Vec<_>>();
+                    .filter_map(|(receiver, method)| {
+                        let method_call = format!(".{}()", method.ident);
+                        if receiver.ends_with(&method_call) {
+                            None  // do not suggest code that is already there (#53348)
+                        } else {
+                            Some(format!("{}{}", receiver, method_call))
+                        }
+                    }) .collect::<Vec<_>>();
                 if !suggestions.is_empty() {
                     err.span_suggestions(expr.span, "try using a conversion method", suggestions);
                 }

--- a/src/test/ui/issues/issue-53348.rs
+++ b/src/test/ui/issues/issue-53348.rs
@@ -1,0 +1,26 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn main() {
+    let mut v = vec!["hello", "this", "is", "a", "test"];
+
+    let v2 = Vec::new();
+
+    v.into_iter().map(|s|s.to_owned()).collect::<Vec<_>>();
+
+    let mut a = String::new();
+    for i in v {
+        a = *i.to_string();
+        //~^ ERROR mismatched types
+        //~| NOTE expected struct `std::string::String`, found str
+        //~| NOTE expected type
+        v2.push(a);
+    }
+}

--- a/src/test/ui/issues/issue-53348.stderr
+++ b/src/test/ui/issues/issue-53348.stderr
@@ -1,0 +1,12 @@
+error[E0308]: mismatched types
+  --> $DIR/issue-53348.rs:20:13
+   |
+LL |         a = *i.to_string();
+   |             ^^^^^^^^^^^^^^ expected struct `std::string::String`, found str
+   |
+   = note: expected type `std::string::String`
+              found type `str`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

